compiler-core/src/language_server/code_action.rs

```diff
@@ -276,3 +276,131 @@ impl<'a> RedundantTupleInCaseSubject<'a> {
         }
     }
 }
+
+// Builder for code action to convert `let assert` into a case expression
+pub struct LetAssertToCase<'a> {
+    module: &'a Module,
+    params: &'a CodeActionParams,
+    actions: Vec<CodeAction>,
+    line_numbers: LineNumbers,
+    pattern_variables: Vec<EcoString>,
+}
+
+impl<'ast> ast::visit::Visit<'ast> for LetAssertToCase<'_> {
+    fn visit_typed_assignment(&mut self, assignment: &'ast ast::TypedAssignment) {
+        let range = src_span_to_lsp_range(assignment.location, &self.line_numbers);
+
+        // Only offer the code action if the cursor is over the statement
+        if !overlaps(range, self.params.range) {
+            return;
+        }
+
+        // This pattern only applies to `let assert`
+        if !matches!(assignment.kind, AssignmentKind::Assert { .. }) {
+            return;
+        };
+
+        // Get the source code for the tested expression
+        let location = assignment.value.location();
+        let expr = self
+            .module
+            .code
+            .get(location.start as usize..location.end as usize)
+            .expect("Location must be valid");
+
+        // Get the source code for the pattern
+        let pattern_location = assignment.pattern.location();
+        let pattern = self
+            .module
+            .code
+            .get(pattern_location.start as usize..pattern_location.end as usize)
+            .expect("Location must be valid");
+
+        let indent = " ".repeat(range.start.character as usize);
+
+        // Figure out which variables are assigned in the pattern
+        self.pattern_variables.clear();
+        self.visit_typed_pattern(&assignment.pattern);
+        let variables = std::mem::take(&mut self.pattern_variables);
+
+        let assigned = match variables.len() {
+            0 => "",
+            1 => variables.first().expect("Variables is length one"),
+            _ => &format!("#({})", variables.join(", ")),
+        };
+
+        let edit = TextEdit {
+            range,
+            new_text: format!(
+                "let {assigned} = case {expr} {{
+{indent}  {pattern} -> {assigned}
+{indent}  _ -> panic
+{indent}}}",
+                // "_" isn't a valid expression, so we just return Nil from the case expression
+                assigned = if assigned == "_" { "Nil" } else { assigned }
```

Comment:
I don't think `_` is ever possible to be used as a variable name, and we're missing the case where no variables are assigned at all, e.g. `let assert [] = x`. Could we have a test for that please